[Parse] Implement "missing 'func' keyword" diagnostic with a fix-it

broadwaylamb opened the pull request:
- When parsing a member list of a type or extension declaration, attempt to parse a function or a property declaration when meeting an identifier, an operator or an opening paren (for tuple declarations).
- Produce the diagnostic with a fix-it suggesting to insert the needed keyword
- Recover parsing as if the declaration with the missing keyword is a function/property declaration

It's not a surprise that this change broke a number of tests.
I've ran smoke tests locally and fixed most of them, however, couldn't figure out how to fix the remaining couple of them (namely [this](https://github.com/apple/swift/blob/0ea82d689053d936e292271c10c9310931cd94e9/test/incrParse/funcs.swift#L25-L27) and [this](https://github.com/apple/swift/blob/0ea82d689053d936e292271c10c9310931cd94e9/test/Driver/working-directory.swift#L66), I failed to understand how's the latter related), so I'd appreciate any advice from incremental parsing people ☺️

Haven't yet run validation tests, wanna deal with the smoke tests first.

When I'm done with the functional changes, I'll run clang-format on the diff.

cc @jrose-apple @DougGregor @dabrahams

Resolves [SR-10477](https://bugs.swift.org/browse/SR-10477).

broadwaylamb committed:
[Parse] Implement "missing 'func' keyword" diagnostic with a fix-it

- When parsing a type or extension declaration, attempt to parse a function or property declaration when meeting an identifier, an operator or a paren (for tuple declarations).
- Produce the diagnostic with a fix-it suggesting to insert the needed keyword
- Recover parsing as if the declaration with the missing keyword is a function/property declaration

Resolves https://bugs.swift.org/browse/SR-10477

jrose-apple commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2936,6 +2946,33 @@ Parser::parseDecl(ParseDeclOptions Flags,
+
+      if ((Tok.isIdentifierOrUnderscore() &&
+           peekToken().isAny(tok::colon, tok::equal, tok::comma)) ||
+          Tok.is(tok::l_paren)) {
```
The l-paren case seems very subtle. Maybe if it were an lparen and the *next* token were an identifier too, but otherwise I'm not sure I trust this as a recovery.

broadwaylamb replied:
At first I thought that that wouldn't catch nested tuples such as `var ((x, y), z) = ((1, 2), 3)`, and I even included it as a test case, but now I see that this is probably very rare and indeed not worth it.

So yeah, I'll change it according to your comment.

jrose-apple commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2936,6 +2946,33 @@ Parser::parseDecl(ParseDeclOptions Flags,
+            .fixItInsert(Tok.getLoc(), "var ");
+        parseLetOrVar(/*HasLetOrVarKeyword=*/false);
+        break;
+      } else if (Tok.isIdentifierOrUnderscore() || Tok.isAnyOperator()) {
```
Nitpick: LLVM/Swift style is to not use `else` after a `break` or `return`.

jrose-apple commented on test/Parse/line-directive.swift:
```diff
@@ -24,6 +24,13 @@ x x // expected-error{{consecutive statements}} {{2-2=;}}
+// expected-error@+5{{operators must have one or two arguments}}
+// expected-error@+4{{member operator '/()' must have at least one argument of type 'S'}}
+// expected-error@+3{{expected '{' in body of function declaration}}
+// expected-error@+2{{consecutive declarations on a line must be separated by ';}}
```
Hrm, this is unfortunate. I guess it's hard to catch the generic case without it though…

jrose-apple commented on test/SourceKit/DocumentStructure/structure.swift.empty.response:
```diff
@@ -1177,7 +1177,40 @@
+        {
+          key.kind: source.lang.swift.decl.function.method.instance,
+          key.accessibility: source.lang.swift.accessibility.internal,
+          key.name: "init(x:y:)",
```
Hm, this seems suspect. An initializer should never be treated as a function!

broadwaylamb replied:
As far as I understand, this is not an initializer.
https://github.com/apple/swift/blob/d7c0a61c9334ae66adf8d48cf87e9441e4d9218f/test/SourceKit/DocumentStructure/Inputs/main.swift#L114-L117

Currently this input produces the "expected declaration" error.

jrose-apple replied:
Ahhh, whoops! Okay, then this is behaving correctly, though I wonder if SourceKit should be escaping it.

broadwaylamb replied:
As far as I can see from the other declarations with backticks, it shouldn't.

Declaration:

https://github.com/apple/swift/blob/7ba9d655dfeff6e6dde58c6ab05453d096f0b58a/test/SourceKit/DocumentStructure/Inputs/main.swift#L120

SourceKit output:

https://github.com/apple/swift/blob/7ba9d655dfeff6e6dde58c6ab05453d096f0b58a/test/SourceKit/DocumentStructure/structure.swift.response#L1193-L1196

jrose-apple replied:
Yeah, but `foo` isn't a reserved word normally. `` func `func` ``, for example, might be more interesting.

Anyway, this isn't changed by your work here, so we don't need to do something about it now.

jrose-apple reviewed (commented):
Thanks for taking this on! Your implementation makes sense, and I think on the whole diagnostics have improved. The main concern I have is about the `init(x:y:)` case in the SourceKit tests. What's causing this patch to treat an initializer name as a regular method?

jrose-apple commented:
I don't see how the Driver test case could be related either. Let's see if it reproduces in our CI.

@swift-ci Please test

jrose-apple commented:
@rintaro, who's the right person to talk to about the [incremental parsing](https://github.com/apple/swift/blob/0ea82d689053d936e292271c10c9310931cd94e9/test/incrParse/funcs.swift#L25-L27) test case? @broadwaylamb, what behavior are you seeing?

jrose-apple commented:
Looks like the Driver failure went away, so it's just the incremental parsing one that matters.

rintaro commented:
@akyrtzi Do you have any idea what's happening?

akyrtzi commented:
@broadwaylamb it's hard to provide any guidance about how to resolve the test failure without seeing how it fails. Could you post the test output ?

broadwaylamb commented:
@akyrtzi here's my output of lit invocation `./llvm/utils/lit/lit.py -sv ${SWIFT_BUILD_DIR}/test-macosx-x86_64/incrParse --filter=funcs.swift`:

https://gist.github.com/broadwaylamb/67cddbc2b9e7b3c379695b73fe899307

There's a lot of diff there, but basically the error is:

```
Test case "REMOVE_FUNC_KEYWORD" of /Users/sergej/Developer/apple/swift/test/incrParse/funcs.swift FAILed
Syntax tree of incremental parsing does not match from-scratch parsing of post-edit file:
```

akyrtzi commented:
I can't see by inspection what the issue is with your changes. My recommendation would be to reduce the test case, find the specific step that differs with your changes vs without, and debug the two code paths. Let us know if you have any questions, apologies for not being more helpful.

broadwaylamb committed:
[incrParse] Fix funcs.swift test case

broadwaylamb commented on test/incrParse/funcs.swift:
```diff
@@ -16,11 +16,11 @@ class InvalidFuncDecls {
   func closingBraceAdded() {
 
-  <<ADD_ClOSING_BRACE|||}>>>
+  <<ADD_ClOSING_BRACE<|||}>>>
```
@akyrtzi looks like this was the typo that caused this test to fail.

akyrtzi replied:
Good catch!

broadwaylamb committed:
[Parse] Simplify missing_func_keyword diagnostic

broadwaylamb commented:
So I've finally found some time and fixed the incrParse test failure, as well as addressed the review comments.

@jrose-apple can I ask you for a second review?

jrose-apple commented:
@swift-ci Please test

jrose-apple commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -216,6 +216,8 @@ ERROR(expected_decl,none,
 ERROR(expected_identifier_in_decl,none,
       "expected identifier in %0 declaration", (StringRef))
+ERROR(expected_keyword_in_decl,none,
+      "expected '%0' keyword in %1 declaration", (StringRef, StringRef))
```
Nitpick: you can use DescriptiveDeclKind for the second parameter here.

broadwaylamb replied:
Okay, will do

broadwaylamb replied:
@jrose-apple can you please review and trigger CI again?

jrose-apple reviewed (commented):
Thanks, it looks good to me! @akyrtzi or @rintaro might have some final comments too.

jrose-apple commented:
@swift-ci Please test macOS

broadwaylamb committed:
Use DescriptiveDeclKind for better expected_keyword_in_decl diagnostic

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
Looks like at least one more thing changed in a test:

```
16:25:58 --- /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/swift/test/SourceKit/DocumentStructure/structure.swift.response	2019-07-05 16:02:54.000000000 -0700
16:25:58 +++ /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/SourceKit/DocumentStructure/Output/structure.swift.tmp.response	2019-07-05 16:25:58.000000000 -0700
16:25:58 @@ -1453,7 +1453,7 @@
16:25:58        key.column: 1,
16:25:58        key.filepath: main.swift,
16:25:58        key.severity: source.diagnostic.severity.error,
16:25:58 -      key.description: "expected 'func' keyword in function declaration",
16:25:58 +      key.description: "expected 'func' keyword in instance method declaration",
16:25:58        key.diagnostic_stage: source.diagnostic.stage.swift.parse,
16:25:58        key.fixits: [
16:25:58          {
16:25:58 
```

jrose-apple commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2982,6 +2992,77 @@ Parser::parseDecl(ParseDeclOptions Flags,
+        DescriptiveDeclKind DescriptiveKind;
+
+        switch (StaticSpelling) {
+          case StaticSpellingKind::None:
```
…and if you're going to do one more push, Swift/LLVM style does not indent cases of a `switch`.

broadwaylamb committed:
Fix failing test cases for expected_keyword_in_decl diagnostic

broadwaylamb commented:
@jrose-apple I hope I’ve fixed the remaining tests and coding style issues. Do I need to clean up the commit history before this is merged?

jrose-apple commented:
Up to you; it's not a requirement. We mostly just care that everything on `master` builds fine, not the PR branches that get merged in.

@swift-ci Please smoke test

broadwaylamb commented:
Okay then, do I need to do anything else?

By the way, any chance this will make it into 5.1? :)

jrose-apple commented:
You can make a PR for the 5.1 branch following the steps at https://swift.org/blog/5-1-release-process/. I know the Playgrounds folks would appreciate it, but it's ultimately up to the branch manager to approve (or one of their delegates).

jrose-apple commented:
And thank you for picking this up!
